src/Oscoin/Protocol/Sync.hs

```diff
@@ -442,3 +451,65 @@ syncBlocks rng = do
          Nothing -> pure () -- TODO(adn) telemetry?
          Just b -> forM_ (scEventHandlers ctx) $ \dispatch ->
                        dispatch (SyncBlock b)
+
+-- | Calls 'isDone' internally, and try syncing blocks if it returns 'False'.
+syncUntil
+    :: ( Monad m
+       , Ord tx
+       , Ord s
+       , Ord (Hash c)
+       , Ord (Beneficiary c)
+       , Buildable (Hash c)
+       )
+    => (SyncContext c tx s m -> Block c tx (Sealed c s) -> Block c tx (Sealed c s) -> Bool)
+    -> SyncT c tx s m ()
+syncUntil finished = do
+    ctx <- ask
+    remoteTip <- getRemoteTip
+    localTip  <- lift (getTip (scLocalChainReader ctx))
+    if finished ctx localTip remoteTip
+       then
+        recordEvent $
+            Telemetry.Events.NodeSyncFinished (blockHash localTip, height localTip)
+       else do
+        recordEvent $
+            Telemetry.Events.NodeSyncStarted (blockHash localTip, height localTip)
+                                             (blockHash remoteTip, height remoteTip)
+        for_ (range localTip remoteTip) syncBlocks
+        newLocalTip <- lift (getTip (scLocalChainReader ctx))
+        recordEvent $
+            Telemetry.Events.NodeSyncFinished (blockHash newLocalTip, height newLocalTip)
```

Comment:
It's hard to make out the actual logic here because of the telemetry - it would be nice if this was factored out.

Reply:
It's a bit of a shame, yes, but if we factor out the telemetry what we lose is a precise callstack of where this happened -- will think about this a bit more.

Reply:
91af9d63